plugins/crypto/src/main/java/org/opensearch/cryptoplugin/CryptoBasePlugin.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+package org.opensearch.cryptoplugin;
+
+import com.amazonaws.encryptionsdk.CryptoAlgorithm;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.client.Client;
+import org.opensearch.cluster.metadata.IndexNameExpressionResolver;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.core.common.io.stream.NamedWriteableRegistry;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.core.xcontent.NamedXContentRegistry;
+import org.opensearch.cryptospi.CryptoKeyProviderExtension;
+import org.opensearch.cryptospi.MasterKeyProvider;
+import org.opensearch.env.Environment;
+import org.opensearch.env.NodeEnvironment;
+import org.opensearch.cryptoplugin.cache.CacheRefresher;
+import org.opensearch.cryptoplugin.cache.MasterKeyCache;
+import org.opensearch.crypto.CryptoManager;
+import org.opensearch.plugins.CryptoPlugin;
+import org.opensearch.plugins.ExtensiblePlugin;
+import org.opensearch.plugins.Plugin;
+import org.opensearch.repositories.RepositoriesService;
+import org.opensearch.script.ScriptService;
+import org.opensearch.threadpool.ExecutorBuilder;
+import org.opensearch.threadpool.FixedExecutorBuilder;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.watcher.ResourceWatcherService;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+/**
+ * AWS plugin for encryption and decryption support.
+ */
+public class CryptoBasePlugin extends Plugin implements CryptoPlugin, ExtensiblePlugin {
```

Comment:
Can we restructure this a bit, to have CryptoBasePlugin sit in core specifically `org.opensearch.plugins` package. Then have specialized implementations as crypto-kms etc as plugin implementations. Wondering do we really need to have `ExtensiblePlugin`?

Reply:
I'm actually not sure I completely follow the structure so let me try to summarize it and please correct me wherever I'm wrong.

- `:server`: Defines a new plugin interface `CryptoPlugin`
- `:lib:opensearch-encryption-sdk`: Defined in #8466. This library looks to be a wrapper/re-implementation of [aws-encryption-sdk-java](https://github.com/aws/aws-encryption-sdk-java/)?
- `::plugins:crypto:spi`: The service provider interface for injecting custom key providers. I'm assuming this is where an AWS KMS implementation could be injected, for example?
- `:plugins:crypto`: This plugin, it wires together the key provider SPI with the `opensearch-encryption-sdk` to implement `CryptoPlugin` used by the core to do the actual encryption/decryption of repository (or any other) data.
- `:lib:opensearch-common`: An interface and a class are added here which are used by several of the above components. 

So there are actually two points of extensibility here: 1) Someone could implement their own `CryptoPlugin` to have total control over encryption and 2) someone can implement `CryptoKeyProviderExtension` in `:plugins:crypto:spi` to provide their own keys but rely on the encryption logic built into `:plugins:crypto`. Is all of this correct? @Bukhtawar Are you suggesting that 2) should be configurable via a plugin but 1) need not be extendable via a plugin?

Also tagging @nknize here as he has been busy de-tangling and modularizing `:server` so would love to get his input.

Reply:
@andrross Structure summarized by you along with their role is absolutely correct. `:libs:opensearch-encryption-sdk` provides extended functionality of partial encryption and decryption on top of what AWS encryption SDK provides.

Reply:
As discussed with @Bukhtawar , this is how we are thinking of structuring crypto : 
1. Move whatever is there in `:plugins:crypto` to `:libs:encryption-sdk` and make lib heavier. Two reasons of doing this :
      1.  When partial encryption/decryption changes are available in `aws-encryption-sdk`, lib becomes leaner and it only consists of code required to manage data keys, providers and wrappers/direct api calls to `aws-encryption-sdk` for actual encryption/decryption.
      2. End users will only have to install one plugin to enable encryption by supplying key provider configs.
2. `:plugins:crypto:spi` will become the new `:plugins:crypto` and its abstractions (key provider abstractions) can now be put where plugin apis are declared. 
3. Build a new `:plugins:crypto-kms` to provide a default implementation of the new `:plugins:crypto`. We can keep building new plugins incrementally for other key providers.

Reply:
Thanks for summarising @vikasvb90. This simplifies a great deal having the end user deal with a single crypto plugin. This also places the heavy lifting logic sit centrally in libs without polluting the core. 
@andrross thoughts?

Reply:
Thanks @vikasvb90, if I understand correctly there are a couple consequences worth calling out:

- The encryption behavior is not pluggable, only the keys. New types of encryption can maybe be added directly to `:libs:encryption-sdk`, but there will be no extension point to add or modify how encryption works.
- The encryption dependencies (`aws-encryption-sdk` and its dependencies) are not optional and will be part of the min distribution and present on the classpath regardless of whether any client-side encryption features are used.

I'm not opposed to this, and I think its okay to be fairly opinionated about encryption as it is something easy to get wrong so having fewer ways to misconfigure is often good. However, I can imagine users needing flexibility here to conform to specific regulatory requirements, for example. I don't want to complicate things based on speculation though, so do you see any decisions here that would be difficult to change if full extensibility became a requirement?

/cc @Bukhtawar @nknize @willyborankin

Reply:
>The encryption dependencies (aws-encryption-sdk and its dependencies) are not optional and will be part of the min distribution and present on the classpath regardless of whether any client-side encryption features are used.

Good point. I feel its the case with any modules today and with security being a first class citizen I am fine keeping this is a core dependency and not making it optional. If you think there is a significant overhead(or tech debt) we can consider it making it a plugin but again that comes with it's own complexity.
I am more inclined in keeping the core encryption logic in `libs` with the required dependency

Reply:
> If you think there is a significant overhead(or tech debt) we can consider it making it a plugin

No, I'm in agreement with you. I think this is the right way to go. (I do think `aws-encryption-sdk` has a slightly unfortunate name in this context because there is nothing AWS-specific about it).

> One point to note is encryption logic isn't commonly proprietary, algorithms used are fairly standard for a vast majority

Agree. Extensibility here would be about supporting the tiny minority to do odd things that we don't want to support in OpenSearch by default. But I think we can cross that bridge if we come to it, and doing the simple thing now is reasonable.